src/main/java/net/fabricmc/installer/UnattendedMain.java

```diff
@@ -0,0 +1,182 @@
+        Path serverLaunchJar = SERVER_DIR.resolve(String.format("fabric-loader-server-%s-minecraft-%s", loaderVersion, gameVersion));
+
+        String mainClass = null;
+
+        if (Files.exists(serverJar) && Files.exists(serverLaunchJar)) {
+            try {
+                mainClass = readMainClass(serverLaunchJar);
+            } catch (IOException e) {
+                // Wont throw here, will try to reinstall
+                System.err.println("Failed to read main class from server launch jar: " + e.getMessage());
+            }
+        }
+
+        if (mainClass != null) {
+            // All seems good, no need to reinstall
+            return new LaunchData(serverJar, serverLaunchJar, mainClass);
+        }
+
+        // TODO improve log output, its quite messy atm
+        Files.createDirectories(SERVER_DIR);
+        ServerInstaller.install(SERVER_DIR, loaderVersion, gameVersion, InstallerProgress.CONSOLE, serverLaunchJar);
+
+        InstallerProgress.CONSOLE.updateProgress(Utils.BUNDLE.getString("progress.download.minecraft"));
+        MinecraftServerDownloader downloader = new MinecraftServerDownloader(gameVersion);
+        downloader.downloadMinecraftServer(serverJar);
+
+        mainClass = readMainClass(serverLaunchJar);
+
+        return new LaunchData(serverJar, serverLaunchJar, mainClass);
+    }
+
+    private static Properties readProperties() throws IOException {
+        Properties properties = new Properties();
+
+        URL config = getConfigFromResources();
+
+        if (config == null) {
+            throw new RuntimeException("Jar does not contain unattended install.properties file");
+        }
+
+        try (InputStream is = config.openStream()) {
+            properties.load(is);
+        } catch (IOException e) {
+            throw new IOException("Failed to read stamped installer data", e);
+        }
+
+        return properties;
+    }
+
+    // Find the mainclass of a jar file
+    private static String readMainClass(Path path) throws IOException {
+        try (ZipFile zipFile = new ZipFile(path.toFile())) {
+            ZipEntry zipEntry = zipFile.getEntry("META-INF/MANIFEST.MF");
+
+            if (zipEntry == null) {
+                throw new IOException("Failed to find manifest in jar");
+            }
+
+            try (InputStream inputStream = zipFile.getInputStream(zipEntry)) {
+                Manifest manifest = new Manifest(inputStream);
+                String mainClass = manifest.getMainAttributes().getValue("Main-Class");
+
+                if (mainClass == null) {
+                    throw new IOException("Jar does not have a Main-Class attribute");
+                }
+
+                return mainClass;
+            }
+        }
+    }
+
+    private static void validateLoaderVersion(String loaderVersion) {
+        int[] versionSplit = Arrays.stream(loaderVersion.split("\\."))
+                .mapToInt(Integer::parseInt)
+                .toArray();
```

Comment:
I'd rather not use stream+lambda to save a few ms for generating+baking the impl class, could just inline the Integer.parseInt below